### PR TITLE
add sentencepiece tokenizer detokenizer

### PR DESCRIPTION
**Pull request by szha**
## Description ##
add sentencepiece tokenizer detokenizer

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [x] Add sentencepiece tokenizer and detokenizer to support loading pre-trained subword models from google/sentencepiece

### CODE REVIEW

**Comment by mli** [comment]
Job PR-380/2 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-380/2/index.html

File: gluonnlp/data/transforms.py [hunk 1]
```diff
@@ -668,3 +668,131 @@ def __call__(self, sample):
...
+    path : str
+        Path to the pre-trained subword tokenization model.
+
+    Examples
+    --------
+    >>> tokenizer = SentencepieceTokenizer('./corpus.full.bpe')
+    >>> detokenizer = SentencepieceDetokenizer('./corpus.full.bpe')
+    >>> sentence = 'This is a very awesome, life-changing sentence.'
+    >>> tokenizer(sentence)
+    ['▁This',
```
Comment (szhengac) [inline gluonnlp/data/transforms.py]: This looks different from the moses BPE, in which subwords are identified by symbols @@. For example, with moses BPE, '_This' will be 'This' and 'ch' will be 'ch@@'.

Comment (szha) [reply gluonnlp/data/transforms.py]: Yes indeed, and I wasn't able to find an option to set the word boundary markers. Is there such option in moses bpe?

Comment (szha) [reply gluonnlp/data/transforms.py]: On the other hand, we may choose to integrate https://github.com/rsennrich/subword-nmt to support both, which carries an MIT license.

**Comment by mli** [comment]
Job PR-380/4 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-380/4/index.html

File: gluonnlp/data/transforms.py (hunk 1 above)
Comment (szha) [reply gluonnlp/data/transforms.py]: I will address this in a separate PR.

**Comment by leezu** [comment]
Thanks @szha. The integration works great.

For the future, I wonder: SentencePiece manages it's own Piece - ID mapping, which we may wan't to use as well. Would it make sense to - in addition to this - introduce a SentencePiece wrapper for our Vocab class? My concern here is that based on a particular tokenization obtained from the tokenizer in this PR, there is no guarantee that all possible tokens the sentencepiece model knows about are contained. So constructing a vocabulary on the gluonnlp side may not make sense. By treating sentencepiece as a vocabulary implementation it seems we may also expose `SampleEncodeAsIds` etc.

**Comment by szha** [comment]
@leezu I think it makes sense to integrate the vocabulary of sentencepiece, but maybe not in the way as a wrapper around the specific functionality. On one hand, the serialized format of vocab in sentencepiece is a simple piece-ID mapping, which can already be covered by the vocab functionality in gluonnlp. On the other hand, if we choose to extend the functionality of our own vocab, we may run into difficulties to make such wrapper compatible with extended interface. Thus, I'd prefer to have the integration to be at the level of serialization mapping rather than wrapper class.

**Comment by szha** [comment]
Actually there's a simple way to get all the tokens without a vocab file.

**Comment by mli** [comment]
Job PR-380/8 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-380/8/index.html

**Comment by mli** [comment]
Job PR-380/10 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-380/10/index.html

**Pull request merged**